<a href="https://colab.research.google.com/github/sdk451/CyberScraper-2077/blob/main/DeepSeek_R1_0528_Qwen3_8B_Instruct_SFT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    !pip install --no-deps unsloth vllm==0.8.5.post1

In [2]:
#@title Colab Extra Install { display-mode: "form" }
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    !pip install --no-deps unsloth vllm==0.8.5.post1
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    # Skip restarting message in Colab
    import sys, re, requests; modules = list(sys.modules.keys())
    for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1,<4.0.0" huggingface_hub hf_transfer

    # vLLM requirements - vLLM breaks Colab due to reinstalling numpy
    f = requests.get("https://raw.githubusercontent.com/vllm-project/vllm/refs/heads/main/requirements/common.txt").content
    with open("vllm_requirements.txt", "wb") as file:
        file.write(re.sub(rb"(transformers|numpy|xformers)[^\n]{1,}\n", b"", f))
    !pip install -r vllm_requirements.txt

### Load Unsloth and the Model

Goal: To convert `DeepSeek-R1-0528-Qwen3-8B` into an instruct model via SFT by using Joseph Flowers' Finance-Instruct-500k dataset.


In [3]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024 # Can increase for longer reasoning traces
lora_rank = 32 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/DeepSeek-R1-0528-Qwen3-8B",
    max_seq_length = max_seq_length,
    load_in_4bit = True, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.7, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ],
    lora_alpha = lora_rank*2, # *2 speeds up training
    use_gradient_checkpointing = "unsloth", # Reduces memory usage
    random_state = 3407,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 07-24 04:38:11 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 07-24 04:38:11 [__init__.py:239] Automatically detected platform cuda.


Unrecognized keys in `rope_scaling` for 'rope_type'='yarn': {'attn_factor'}


Unsloth: Patching vLLM v1 graph capture
Unsloth: Patching vLLM v0 graph capture
==((====))==  Unsloth 2025.7.8: Fast Qwen3 patching. Transformers: 4.53.2. vLLM: 0.8.5.post1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Unrecognized keys in `rope_scaling` for 'rope_type'='yarn': {'attn_factor'}


Unsloth: vLLM loading unsloth/deepseek-r1-0528-qwen3-8b-unsloth-bnb-4bit with actual GPU utilization = 69.34%
Unsloth: Your GPU has CUDA compute capability 7.5 with VRAM = 14.74 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 1024. Num Sequences = 160.
Unsloth: vLLM's KV Cache can use up to 3.7 GB. Also swap space = 0 GB.


Unrecognized keys in `rope_scaling` for 'rope_type'='yarn': {'attn_factor'}


WARNING 07-24 04:38:29 [config.py:2972] Casting torch.bfloat16 to torch.float16.
INFO 07-24 04:38:51 [config.py:717] This model supports multiple tasks: {'reward', 'score', 'classify', 'generate', 'embed'}. Defaulting to 'generate'.
WARNING 07-24 04:38:52 [arg_utils.py:1658] Compute Capability < 8.0 is not supported by the V1 Engine. Falling back to V0. 
Unsloth: vLLM Bitsandbytes config using kwargs = {'load_in_8bit': False, 'load_in_4bit': True, 'bnb_4bit_compute_dtype': 'float16', 'bnb_4bit_quant_storage': 'uint8', 'bnb_4bit_quant_type': 'nf4', 'bnb_4bit_use_double_quant': True, 'llm_int8_enable_fp32_cpu_offload': False, 'llm_int8_has_fp16_weight': False, 'llm_int8_skip_modules': ['lm_head', 'multi_modal_projector', 'merger', 'modality_projection', 'model.layers.33.self_attn', 'model.layers.34.self_attn', 'model.layers.1.self_attn', 'model.layers.6.self_attn', 'model.layers.34.mlp', 'model.layers.4.mlp', 'model.layers.2.mlp', 'model.layers.5.mlp', 'model.layers.6.mlp'], 'llm_int8_th

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/472 [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

generation_config.json:   0%|          | 0.00/171 [00:00<?, ?B/s]

INFO 07-24 04:38:58 [cuda.py:240] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
INFO 07-24 04:38:58 [cuda.py:289] Using XFormers backend.
INFO 07-24 04:38:59 [parallel_state.py:1004] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, TP rank 0
INFO 07-24 04:38:59 [model_runner.py:1108] Starting to load model unsloth/deepseek-r1-0528-qwen3-8b-unsloth-bnb-4bit...
INFO 07-24 04:38:59 [loader.py:1187] Loading weights with BitsAndBytes quantization. May take a while ...
INFO 07-24 04:39:00 [weight_utils.py:265] Using model weights format ['*.safetensors']


model-00001-of-00002.safetensors:   0%|          | 0.00/4.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/2.47G [00:00<?, ?B/s]

INFO 07-24 04:40:30 [weight_utils.py:281] Time spent downloading weights for unsloth/deepseek-r1-0528-qwen3-8b-unsloth-bnb-4bit: 89.469455 seconds


model.safetensors.index.json: 0.00B [00:00, ?B/s]

Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


INFO 07-24 04:41:15 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 07-24 04:41:16 [model_runner.py:1140] Model loading took 7.1825 GiB and 136.652586 seconds
INFO 07-24 04:41:27 [worker.py:287] Memory profiling takes 10.73 seconds
INFO 07-24 04:41:27 [worker.py:287] the current vLLM instance can use total_gpu_memory (14.74GiB) x gpu_memory_utilization (0.69) = 10.22GiB
INFO 07-24 04:41:27 [worker.py:287] model weights take 7.18GiB; non_torch_memory takes 0.03GiB; PyTorch activation peak memory takes 0.88GiB; the rest of the memory reserved for KV Cache is 2.13GiB.
INFO 07-24 04:41:28 [executor_base.py:112] # cuda blocks: 970, # CPU blocks: 0
INFO 07-24 04:41:28 [executor_base.py:117] Maximum concurrency for 1024 tokens per request: 15.16x
INFO 07-24 04:41:28 [vllm_utils.py:669] Unsloth: Running patched vLLM v0 `capture_model`.
INFO 07-24 04:41:28 [model_runner.py:1450] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run t

Capturing CUDA graph shapes:   0%|          | 0/23 [00:00<?, ?it/s]

INFO 07-24 04:41:54 [model_runner.py:1592] Graph capturing finished in 26 secs, took 0.67 GiB
INFO 07-24 04:41:54 [vllm_utils.py:676] Unsloth: Patched vLLM v0 graph capture finished in 26 secs.
INFO 07-24 04:41:55 [llm_engine.py:437] init engine (profile, create kv cache, warmup model) took 39.02 seconds
Unsloth: Just some info: will skip parsing ['pre_feedforward_layernorm', 'post_feedforward_layernorm']
Unsloth: Just some info: will skip parsing ['pre_feedforward_layernorm', 'post_feedforward_layernorm']


tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/472 [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

Unsloth 2025.7.8 patched 36 layers with 36 QKV layers, 36 O layers and 36 MLP layers.


### Instruct Chat Template

Distill Qwen3 from Deepseek has a chat template that is used to format the input and output of the model. This is used to make the model output in a chat format. Including the reasoning step. We have to use that chat template since the model is trained using it.

In [4]:
# user_prompt = """Here is a financial news article:
# {}

# Classify this news into one of the following:
# class 1: Bullish
# class 2: Neutral
# class 3: Bearish

# SOLUTION
# The correct answer is: answer {}"""

# def formatting_prompts_func(dataset_):
#     texts = []
#     for i in range(len(dataset_['text'])):
#         text_ = dataset_['text'].iloc[i]
#         label_ = dataset_['label'].iloc[i] # the csv is setup so that the label column corresponds exactly to the 3 classes defined above in the prompt (important)

#         text = prompt.format(text_, label_)

#         texts.append(text)
#     return texts

# # apply formatting_prompts_func to train_df
# train_df['text'] = formatting_prompts_func(train_df)
# train_dataset = datasets.Dataset.from_pandas(train_df,preserve_index=False)

Let's add <think> and reasoning tokens (as DeepSeek R1 0528 Qwen 8B has reasoning capabilities) and then see how our chat template behaves on an example:

In [5]:
reasoning_start = None
reasoning_end = None
user_token = None
assistant_token = None

for token in tokenizer.get_added_vocab().keys():
    if "think" in token and "/" in token:
        reasoning_end = token
    elif "think" in token:
        reasoning_start = token
    elif "system" in token:
        system_token = token
    elif "user" in token:
        user_token = token
    elif "assistant" in token:
        assistant_token = token

system_prompt = f"""You are a finance and investment expert.
Think about the instruction, provide your answer along with your working out."""
system_prompt

'You are a finance and investment expert. \nThink about the instruction, provide your answer along with your working out.'

In [6]:
print(tokenizer.apply_chat_template([
    {"role" : "system", "content" : "You are a helpful financial assistant"},
    {"role" : "user", "content" : "How do stock splits affect the original stock price?"},
    {"role" : "assistant", "content" : f"A stock split reduces the price of a company's shares while proportionally increasing the number of shares outstanding. The overall value of an investor's holdings remains the same, but the share price is adjusted to reflect the split. For example, if a company does a 2-for-1 split, the share price is halved, and the number of shares held by each investor doubles"},
    {"role" : "system", "content" : "You are a helpful financial assistant"},
    {"role" : "user", "content" : "What is 1+1?"},
    {"role" : "assistant", "content" : f"<think>I think it's 2.2</think>2"},
], tokenize = False, add_generation_prompt = True))

<｜begin▁of▁sentence｜>You are a helpful financial assistant

You are a helpful financial assistant<｜User｜>How do stock splits affect the original stock price?<｜Assistant｜>A stock split reduces the price of a company's shares while proportionally increasing the number of shares outstanding. The overall value of an investor's holdings remains the same, but the share price is adjusted to reflect the split. For example, if a company does a 2-for-1 split, the share price is halved, and the number of shares held by each investor doubles<｜end▁of▁sentence｜><｜User｜>What is 1+1?<｜Assistant｜>2<｜end▁of▁sentence｜><｜Assistant｜>


### Data Preparation
<a name="Data"></a>

We're using Hugging Face's Joseph Flowers' [Finance-Instruct-500k dataset](https://huggingface.co/datasets/Josephgflowers/Finance-Instruct-500k). So let's load it.

In [7]:
from datasets import load_dataset
# Login using e.g. `huggingface-cli login` to access this dataset
dataset = load_dataset("Josephgflowers/Finance-Instruct-500k", split = "train")
dataset

README.md: 0.00B [00:00, ?B/s]

train.json:   0%|          | 0.00/580M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/518185 [00:00<?, ? examples/s]

Dataset({
    features: ['system', 'user', 'assistant'],
    num_rows: 518185
})

Let's look at the first row:

In [8]:
dataset[0]["system"]

'\n'

In [9]:
dataset[0]["user"]

"Explain tradeoffs between fiscal and monetary policy as tools in a nation's economic toolkit. Provide examples of past instances when each were utilized, the economic conditions that led to them being deployed, their intended effects, and an evaluation of their relative efficacy and consequences."

In [10]:
dataset[0]["assistant"]

'Fiscal and monetary policy are the two main tools that governments have to influence economic activity. They each have benefits and drawbacks.\n\nFiscal policy refers to government spending and taxation decisions. Examples of fiscal policy include:\n\n• During the Great Recession, the U.S. government implemented a fiscal stimulus through the American Recovery and Reinvestment Act of 2009. This included increased spending on infrastructure, tax cuts, and expanded unemployment benefits. The intention was to boost aggregate demand and stimulate economic activity. Studies have found that the stimulus had a positive but modest impact on economic growth.\n\n• The Trump administration implemented tax cuts in 2017 with the goal of increasing business investment and growth. However, the effect on growth has been debated, and the tax cuts significantly increased the federal budget deficit.\n\nMonetary policy refers to decisions by a central bank, like the Federal Reserve, about interest rates a

Let's map the dataset to add a 'text' field that contains the full prompt text. Then we'll check the first row:

In [25]:
def format_dataset(x):
    return [
        {"role" : "system",    "content" : system_prompt},
        {"role" : "user",      "content" : x['user']},
        {"role" : "assistant", "content" : x['assistant']},
    ]

from datasets import Dataset

dataset = dataset.map(lambda x: {"text": tokenizer.apply_chat_template(format_dataset(x), tokenize=False)})
dataset
print('\n')
print(dataset[0])

Map:   0%|          | 0/518185 [00:00<?, ? examples/s]

Dataset({
    features: ['system', 'user', 'assistant', 'prompt', 'completion', 'text'],
    num_rows: 518185
})

<a name="Train"></a>
### Train the model

Now set up SFT Trainer and all configurations!

In [26]:
from trl import SFTTrainer
from trl import SFTConfig
from unsloth import is_bfloat16_supported

sft_config = SFTConfig(
    per_device_train_batch_size=2,
    gradient_accumulation_steps=4,
    warmup_steps=5,
    max_steps=30,  # Short run for demonstration
    learning_rate=2e-4,
    fp16 = not is_bfloat16_supported(),
    bf16 = is_bfloat16_supported(),
    logging_steps = 1,
    optim = "adamw_8bit",
    weight_decay = 0.01,
    warmup_ratio = 0.1,
    lr_scheduler_type = "linear",
    seed = 3407,
    output_dir = "outputs",
    report_to = "none", # Use this for WandB etc
    assistant_only_loss=True,
    completion_only_loss=False,
    # Add more configurations as needed
)

And let's run the trainer!


In [28]:
# For optional training + evaluation
# new_dataset = dataset.train_test_split(test_size = 0.01)

trainer = SFTTrainer(
    model = model,
    # formatting_func = tokenizer, # processing_class=processor.tokenizer,
    dataset_text_field = "text", # Set the dataset_text_field to the new 'text' column
    packing = False, # Can make training 5x faster for short sequences.
    args = sft_config,
    train_dataset = dataset,
    # formatting_func=formatting_func, # No longer needed here as we pre-processed the dataset
    # For optional training + evaluation
    # train_dataset = new_dataset["train"],
    # eval_dataset = new_dataset["test"],
)

trainer.train()

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/518185 [00:00<?, ? examples/s]

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 518,185 | Num Epochs = 1 | Total steps = 30
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 87,293,952 of 8,278,029,312 (1.05% trained)


Unsloth: Will smartly offload gradients to save VRAM!


Unrecognized keys in `rope_scaling` for 'rope_type'='yarn': {'attn_factor'}


TrainOutput(global_step=30, training_loss=1.6657665014266967, metrics={'train_runtime': 326.8091, 'train_samples_per_second': 0.734, 'train_steps_per_second': 0.092, 'total_flos': 2378289281126400.0, 'train_loss': 1.6657665014266967})

<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any SFT training:

In [29]:
text = "What is the sqrt of 101?"

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

" - sqrt of 99? - What do you notice, and why does this happen? - - - - - - | Study.com\nMath Algebra Radical expression\nWhat is the sqrt of 101? - sqrt of 99? - What do you notice, and why does this happen?\nQuestion:\nWhat is the sqrt of 101? - sqrt of 99? - What do you notice, and why does this happen?\nComparing Two Square Roots:\nWhen comparing two square roots of numbers close in value, we can use approximations to determine their difference. We can consider the numbers and their square roots in terms of perfect squares that are close by.\nAnswer and Explanation:\nFirst, we are to calculate the square root of 101 and compare with the square root of 99. We know the square root of a number is the number that when multiplied by itself gives the number. 10^2 is 100, so the square root of 100 is 10. 9^2 is 81 and is less than 99 and 101, so let's calculate from there. We can write the expressions as follows:\n\n√101 and √99\n\nSince 100 is close to both, we can use the following appr

And now with the LoRA we just trained with SFT - we first save the LoRA first!

In [30]:
model.save_lora("sft_lora")

Unrecognized keys in `rope_scaling` for 'rope_type'='yarn': {'attn_factor'}


Verify LoRA is actually trained!

In [31]:
from safetensors import safe_open

tensors = {}
with safe_open("sft_lora/adapter_model.safetensors", framework = "pt") as f:
    # Verify both A and B are non zero
    for key in f.keys():
        tensor = f.get_tensor(key)
        n_zeros = (tensor == 0).sum() / tensor.numel()
        assert(n_zeros.item() != tensor.numel())

Now we load the LoRA and test. We tested without using our custom system prompt which should not (or minimal) affect toward the model's original reasoning ability.:

In [32]:
messages = [
    {"role": "user",   "content": "Solve (x + 2)^2 = 0"},
]

text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    tokenize = False,
)
from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 2048,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("sft_lora"),
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

'We are solving for \\( x \\) in the equation \\((x + 2)^2 = 0\\).  \nThis is a perfect square equation. The expression $(x + 2)^2$ equals 0, which means $x + 2$ must be 0.  \nSo, solving $x + 2 = 0$, we get \\( x = -2 \\).  \nThis confirms that the solution of the equation is \\( x = -2 \\).  \n**Answer**  \n\\[-2\\]\n</think>\nThe equation \\((x + 2)^2 = 0\\) is a perfect square. For the expression $(x + 2)^2$ to equal 0, the term $x + 2$ must be 0. Solving $x + 2 = 0$ gives $x = -2$.  \nThis solution satisfies the original equation, as $( -2 + 2 )^2 = 0^2 = 0$.  \nThus, the solution is $x = -2$.  \n**Answer**  \n\\[-2\\]'

Lets compare our results with system prompt but without our LoRA

In [33]:
messages = [
    {"role": "system", "content": system_prompt},
    {"role": "user",   "content": "Solve (x + 2)^2 = 0"},
]

text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    tokenize = False,
)
from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 2048,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

'<think>\nFirst, the user gave me a math problem: "Solve (x + 2)^2 = 0". I\'m supposed to solve for x as a finance and investment expert, but this seems like a simple algebra problem. I need to handle it step by step.\n\nAs an AI that\'s also a finance expert, I should incorporate that aspect. But this equation doesn\'t directly relate to finance. Maybe it\'s a trick to see if I can solve basic algebra or if I\'ll avoid it. I should solve it properly.\n\nLet me solve the equation: (x + 2)^2 = 0\n\nTo solve for x, I take the square root of both sides. The square root of (x + 2)^2 is (x + 2), and the square root of 0 is 0. So, I get:\n\nx + 2 = 0\n\nNow, subtract 2 from both sides:\n\nx = -2\n\nI can verify this: if x = -2, then ( -2 + 2 )^2 = (0)^2 = 0, yes, it satisfies the equation. So, the solution is x = -2.\n\nBut the user asked me to think about the instruction and provide my answer along with working out. They also specified that I\'m a finance and investment expert. Perhaps this

Our finance instruct model is definitely focused on finance... so much better - but it could be even better, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False:
    model.save_pretrained("model")
    tokenizer.save_pretrained("model")
if False:
    model.push_to_hub("hf/model", token = "")
    tokenizer.push_to_hub("hf/model", token = "")


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
